src/main/java/com/amazon/opendistroforelasticsearch/search/asynchronous/service/AsynchronousSearchService.java

```diff
@@ -370,21 +366,52 @@ private void cancelAndFreeActiveAndPersistedContext(AsynchronousSearchActiveCont
         asynchronousSearchContext.acquireContextPermitIfRequired(wrap(
                 releasable -> {
                     releasableReference.set(releasable);
-                    boolean response = freeActiveContext(asynchronousSearchContext);
                     if (asynchronousSearchContext.keepOnCompletion()) {
-                        cancelTask(asynchronousSearchContext, cancelTaskReason, () -> groupedDeletionListener.onResponse(response));
+                        //TODO extract cancel task logic to a function to make it unit-testable
+                        cancelTask(asynchronousSearchContext, cancelTaskReason, wrap(cancelTasksResponse -> {
```

Comment:
Instead of passing a `noCancellationRunnable` should we have `cancel` return boolean and take care of handing grouped listeners in the caller